File: 0130. 被围绕的区域/dfs.cpp
// 28ms
class Solution {
public:
    int row, col;
    vector<vector<bool>> mark;

    void solve(vector<vector<char>>& board) {
        row = board.size();
        if (row == 0) {
            return;
        }
        col = board[0].size();
        mark = vector(row, vector<bool>(col, false));
        for (int i = 0; i < col; i += 1) {
            if (board[0][i] == 'O') {
                dfs(0, i, board);
            }
            if (board[row - 1][i] == 'O') {
                dfs(row - 1, i, board);
            }
        }
        for (int i = 0; i < row; i += 1) {
            if (board[i][0] == 'O') {
                dfs(i, 0, board);
            }
            if (board[i][col - 1] == 'O') {
                dfs(i, col - 1, board);
            }
        }
        for (int i = 0; i < row; i += 1) {
            for (int j = 0; j < col; j += 1) {
                if (board[i][j] == 'O' && !mark[i][j]) {
                    board[i][j] = 'X';
                }
            }
        }
    }

    void dfs(int x, int y, vector<vector<char>>& board) {
        if (board[x][y] == 'O' && !mark[x][y]) {
            mark[x][y] = true;
            if (x - 1 >= 0) {
                dfs(x - 1, y, board);
            }
            if (x + 1 < row) {
                dfs(x + 1, y, board);
            }
            if (y - 1 >= 0) {
                dfs(x, y - 1, board);
            }
            if (y + 1 < col) {
                dfs(x, y + 1, board);
            }
        }
    }
};

File: 0079. 单词搜索/DFS.js
// 128ms
var exist = function(board, word) {
    var table = board.map(i => [])
    const pos = [[-1, 0], [1, 0], [0, -1], [0, 1]]
    const len = word.length, row = board.length, col = board[0].length

    function search(x, y, index) {
        if (board[x][y] == word[index]) {
            index += 1
            if (index == len) {
                return true
            } else {
                let result = false
                table[x][y] = 1
                for (let i = 0; i < 4; i += 1) {
                    let px = x + pos[i][0], py = y + pos[i][1]
                    if (px >= 0 && px < row && py >= 0 && py < col && table[px][py] != 1) {
                        result = result || search(px, py, index)
                    }
                }
                table[x][y] = undefined
                return result
            }
        } else {
            return false
        }
    }
    
    for (let i = 0; i < row; i += 1) {
        for (let j = 0; j < col; j += 1) {
            if (search(i, j, 0)) {
                return true
            }
        }
    }
    return false
};